library/psa_crypto_driver_wrappers.c

```diff
@@ -256,23 +256,49 @@ static psa_status_t get_expected_key_size( const psa_key_attributes_t *attribute
                                            size_t *expected_size )
 {
     size_t buffer_size = 0;
-    if( PSA_KEY_LIFETIME_GET_LOCATION( attributes->core.lifetime ) == PSA_KEY_LOCATION_LOCAL_STORAGE )
+    psa_key_location_t location = PSA_KEY_LIFETIME_GET_LOCATION(attributes->core.lifetime);
+    psa_key_type_t key_type = attributes->core.type;
+    size_t key_bits = attributes->core.bits;
+
+    switch( location )
     {
-        buffer_size = PSA_KEY_EXPORT_MAX_SIZE( attributes->core.type,
-                                               attributes->core.bits );
+        case PSA_KEY_LOCATION_LOCAL_STORAGE:
+            buffer_size = PSA_KEY_EXPORT_MAX_SIZE( key_type,
+                                                   key_bits );
 
-        if( buffer_size == 0 )
-            return( PSA_ERROR_NOT_SUPPORTED );
+            if( buffer_size == 0 )
+                return( PSA_ERROR_NOT_SUPPORTED );
 
-        *expected_size = buffer_size;
-        return( PSA_SUCCESS );
-    }
-    else
-    {
-        /* TBD: opaque driver support: need to calculate size through a
-         * driver-defined size function, since the size of an opaque (wrapped)
-         * key will be different for each implementation. */
-        return( PSA_ERROR_NOT_SUPPORTED );
+            *expected_size = buffer_size;
+            return( PSA_SUCCESS );
+
+#if defined(PSA_CRYPTO_DRIVER_TEST)
+        case PSA_CRYPTO_TEST_DRIVER_LIFETIME:
+            /* TBD: opaque driver support: need to calculate size through a
+             * driver-defined size function, since the size of an opaque (wrapped)
+             * key will be different for each implementation. */
+            if( TEST_KEY_CONTEXT_USE_SIZE_FUNCTION )
+            {
+                *expected_size = test_size_function( key_type,
+                                                     key_bits );
+            }
+            else if( PSA_KEY_TYPE_IS_KEY_PAIR( key_type ) )
```

Comment:
I am confused here. Why is this test code (line 285 to 292) not in `test_size_function()` and thus `test_size_function()` not just returning 0?

Otherwise if for the driver the `size_function()` is not implemented then it is the job of this "eventually generated" C module (driver wrapper) to compute the size from the information in the JSON file I think. Thus I would expect this backup code here not to be test code as it is now. Does it make sense?

Reply:
@ronald-cron-arm I think your assumption about the size_function and having it auto generate code is correct. I was under the impression that eventually the psa_crypto_driver_wrapper.c would be auto generated and this task is simply helping figure out what the auto generated code will eventually need to look like by having the code I added for this pull request. Does that explanation help? Do I need to make further changes?

Reply:
ok thus I understand the code between 285 and 296 as:
the base_size, public_key_size and symmetric factor properties of the key_context property of the test driver would be defined as respectively TEST_KEY_CONTEXT_BASE_SIZE, TEST_KEY_CONTEXT_PUBLIC_KEY_SIZE and TEST_KEY_CONTEXT_SYMMETRIC_FACTOR which are C mnemonics defined in a test driver header (namely test/drivers/size.h). This makes sense to me. Just one question: why public_key_size and store_public_key don't appear here?

Otherwise I am not sure why test_size_function does not do the same calculations as here?

Reply:
From my understand the test_size_function() allows the user to customize the storage size for the key based on the customer driver needs they may have. If the driver does not need the customization they simply will not use it. The current return value of 0 is temporary and needs to be updated along with other parts for validating that using a custom size will perform the required operations correctly. I do not believe that is in scope for this work. If that is not correct then that needs to be updated to some other value, and a new test needs to be added to make sure that is working properly. 

The public_key_size is in the section you were asking about on line 287. The option for store_public_key will need to go in the other opaque functions that will be doing the actual storing of the key and need to account for that setting. I think those functions are still in flight so those changes are not part of this pull request.

I'm definitely not the expert on why the design was put this way, I'm simply trying to implement what was designed to make progress with this feature.